pints/toy/_toy_classes.py

```diff
@@ -61,3 +64,155 @@ def suggested_times(self):
+
+    Models implementing :meth:`_rhs()`, :meth:`jacobian()` and :meth:`_dfdp()`
+    can be used to create a :class:`pints.ForwardModelS1`.
+    """
+    def _dfdp(self, y, t, p):
+        """
+        Returns the derivative of the ODE RHS at time ``t``, with respect to
+        model parameters ``p``.
+
+        Parameters
+        ----------
+        y
+            The state vector at time ``t`` (with length ``n_outputs``).
+        t
+            The time to evaluate at (as a scalar).
+        p
+            A vector of model parameters (of length ``n_parameters``).
+
+        Returns
+        -------
+        A matrix of dimensions ``n_parameters`` by ``n_parameters``.
+        """
+        raise NotImplementedError
+
+    def jacobian(self, y, t, p):
+        """
+        Returns the Jacobian (the derivative of the RHS ODE with respect to the
+        outputs) at time ``t``.
+
+        Parameters
+        ----------
+        y
+            The state vector at time ``t`` (with length ``n_outputs``).
+        t
+            The time to evaluate at (as a scalar).
+        p
+            A vector of model parameters (of length ``n_parameters``).
+
+        Returns
+        -------
+        A matrix of dimensions ``n_outputs`` by ``n_outputs``.
+        """
+        raise NotImplementedError
+
+    def _rhs(self, y, t, p):
+        """
+        Returns the evaluated RHS (``dy/dt``) for a given state vector ``y``,
+        time ``t``, and parameter vector ``p``.
+
+        Parameters
+        ----------
+        y
+            The state vector at time ``t`` (with length ``n_outputs``).
+        t
+            The time to evaluate at (as a scalar).
+        p
+            A vector of model parameters (of length ``n_parameters``).
+
+        Returns
+        -------
+        A vector of length ``n_outputs``.
+        """
+        raise NotImplementedError
+
+    def _rhs_S1(self, y_and_dydp, t, p):
+        """
+        Forms the RHS of ODE for numerical integration to obtain both outputs
+        and sensitivities.
+
+        Parameters
+        ----------
+        y_and_dydp
+            A combined vector of states (elements ``0`` to ``n_outputs - 1``)
+            and sensitivities (elements ``n_outputs`` onwards).
+        t
+            The time to evaluate at (as a scalar).
+        p
+            A vector of model parameters (of length ``n_parameters``).
+
+        Returns
+        -------
+        A vector of length ``n_outputs + n_parameters``.
+        """
+        y = y_and_dydp[0:self.n_outputs()]
+        dydp = y_and_dydp[self.n_outputs():].reshape((self.n_parameters(),
+                                                      self.n_outputs()))
+        dydt = self._rhs(y, t, p)
+        d_dydp_dt = (
+            np.matmul(dydp, np.transpose(self.jacobian(y, t, p))) +
+            np.transpose(self._dfdp(y, t, p)))
+        return np.concatenate((dydt, d_dydp_dt.reshape(-1)))
+
+    def simulate(self, parameters, times):
```

Comment:
This one at least is functionally necessary...
@martinjrobins what do you think?

At the moment we've got the models themselves extending both ToyODEModel and either ForwardModel or ForwardModelS1 (they get to choose, which is important!)

With ToyModel we can avoid having a simulate(), but for ToyODEModel it's kind of necessary that we do. I think the cleanest solution would be to split ToyODEModel into a derivative-free and S1 class, and have those extend ForwardModel and ForwardModelS1. But what we have now works just as well and is probably easier...

Reply:
I think let's avoid splitting things up as it's already quite a complicated class ecosystem.